### PR TITLE
[CBRD-24387] Rewrite page handling for overflow oid in btree

### PR DESCRIPTION
**Pull request by ctshim**
http://jira.cubrid.org/browse/CBRD-24387

* Perform unnecessary code cleanup and rewrite without changing functionality.

- The btree_rv_save_keyval_for_undo_two_objects() is called with BTREE_OVERFLOW_NODE only.
- Fix BTREE_INIT_MVCC_HEADER()
- Fix heap_attrinfo_generate_key()
   When func_index_info is set, information necessary for single key generation is read from this function.
- Change space_slot_size() and space_header_size() from a function to a macro.
- The following functions were subdivided.
    xbtree_load_index(), btree_construct_leafs(), online_index_builder(), xbtree_load_online_index()
- Remove scan_cache argument from heap_attrinfo_read_dbvalues()

### CODE REVIEW

File: src/storage/slotted_page.h [hunk 1]
```diff
@@ -92,8 +92,10 @@ struct spage_slot
 extern void spage_boot (THREAD_ENTRY * thread_p);
 extern void spage_finalize (THREAD_ENTRY * thread_p);
 extern void spage_free_saved_spaces (THREAD_ENTRY * thread_p, void *first_save_entry);
-extern int spage_slot_size (void);
-extern int spage_header_size (void);
+
+#define spage_slot_size()   (sizeof(SPAGE_SLOT))
+#define spage_header_size() (sizeof(SPAGE_HEADER))
```
Comment (hgryoo) [inline src/storage/slotted_page.h]: Please move them over to the function declarations

File: src/storage/page_buffer.h [hunk 2]
```diff
@@ -60,6 +60,13 @@ extern const VPID vpid_Null_vpid;
 #define PGBUF_PAGE_MODIFY_ARGS(pg, prev_lsa) \
   PGBUF_PAGE_VPID_AS_ARGS (pg), LSA_AS_ARGS (prev_lsa), PGBUF_PAGE_LSA_AS_ARGS (pg)
 
+
+#define pgbuf_unfix_and_init_after_check(thread_p, pgptr) \
+  if((pgptr)) { \
+    pgbuf_unfix ((thread_p), (pgptr)); \
+    (pgptr) = NULL; \
+  }
```
Comment (hgryoo) [inline src/storage/page_buffer.h]:
```suggestion
#define pgbuf_unfix_and_init_after_check(thread_p, pgptr) \
  do  {  \
    if((pgptr))  \
      {  \
        pgbuf_unfix ((thread_p), (pgptr)); \
        (pgptr) = NULL; \
      }  \
  } while (0)
```

File: src/storage/heap_file.c [hunk 3]
```diff
@@ -12715,6 +12715,15 @@ heap_attrinfo_generate_key (THREAD_ENTRY * thread_p, int n_atts, int *att_ids, i
 
   if (func_index_info)
     {
+      if (func_index_info->expr)
+	{
+	  if (heap_attrinfo_read_dbvalues (thread_p, cur_oid, recdes, NULL,
+					   func_index_info->expr->cache_attrinfo) != NO_ERROR)
```
Comment (beyondykk9) [inline src/storage/heap_file.c]: the function **heap_attrinfo_read_dbvalues** doesn't use **scan_cache** which is an argument by calling the function. the function **heap_attrinfo_read_dbvalues** needs to fix to exclude **scan_cache** argument.

Comment (beyondykk) [reply src/storage/heap_file.c:12721]: the cur_oid is not used after this routine.

File: src/storage/btree_load.c [hunk 4]
```diff
@@ -264,6 +260,40 @@ static int online_index_builder (THREAD_ENTRY * thread_p, BTID_INT * btid_int, H
...
+  MVCC_REC_HEADER mvcc_header;
+
+  DB_VALUE this_key;		/* Key value in this sorted item (specified with in_recdes) */
+  BTREE_MVCC_INFO mvcc_info;
+  bool is_btree_ops_log;
+
+  OID orig_oid;
+  OID orig_class_oid;
+  MVCC_REC_HEADER orig_mvcc_header;
+} S_PARAM_ST;
```
Comment (beyondykk9) [inline src/storage/btree_load.c:276]: what does mean S_PARAM_ST? sort parameter structure?

File: src/storage/btree_load.c [hunk 5]
```diff
@@ -2211,624 +2226,781 @@ btree_first_oid (THREAD_ENTRY * thread_p, DB_VALUE * this_key, OID * class_oid,
...
+  if (sort_args->key_type->type->data_writeval (&buf, dbvalue_ptr) != NO_ERROR)
+    {
+      return ER_FAILED;
+    }
+
+  recdes->length = CAST_STRLEN (buf.ptr - buf.buffer);
+  return NO_ERROR;
+}
+
+static int
```
Comment (beyondykk9) [inline src/storage/btree_load.c:2408]: please put a comment as function header to explain the function.

File: src/storage/btree_load.c [hunk 6]
```diff
@@ -264,6 +260,40 @@ static int online_index_builder (THREAD_ENTRY * thread_p, BTID_INT * btid_int, H
...
+
+  DB_VALUE this_key;		/* Key value in this sorted item (specified with in_recdes) */
+  BTREE_MVCC_INFO mvcc_info;
+  bool is_btree_ops_log;
+
+  OID orig_oid;
+  OID orig_class_oid;
+  MVCC_REC_HEADER orig_mvcc_header;
+} S_PARAM_ST;
+
```
Comment (beyondykk9) [inline src/storage/btree_load.c:277]: the below functions is related to "btree load", so it looks bettter of the name to start with "btree_load" or "bt_load"

File: src/storage/btree_load.c [hunk 7]
```diff
@@ -264,6 +260,40 @@ static int online_index_builder (THREAD_ENTRY * thread_p, BTID_INT * btid_int, H
...
+static int put_buf_to_record (RECDES * recdes, SORT_ARGS * sort_args, int value_has_null, int key_len,
+			      OID * prev_oid, MVCC_REC_HEADER * mvcc_header, DB_VALUE * dbvalue_ptr, int cur_class,
+			      bool is_btree_ops_log);
+static int get_buf_from_record (OR_BUF * buf, RECDES * recdes, LOAD_ARGS * load_args, S_PARAM_ST * pparam, bool copy);
+static int first_leaf_page_and_init_args (THREAD_ENTRY * thread_p, LOAD_ARGS * load_args, S_PARAM_ST * pparam);
+static int make_new_record_on_leaf_page (THREAD_ENTRY * thread_p, LOAD_ARGS * load_args, S_PARAM_ST * pparam,
+					 int *sp_success);
+static int mvcc_delid_invalid (THREAD_ENTRY * thread_p, LOAD_ARGS * load_args, S_PARAM_ST * pparam);
```
Comment (beyondykk9) [inline src/storage/btree_load.c]:
```suggestion
static int invalidate_mvcc_del_id (THREAD_ENTRY * thread_p, LOAD_ARGS * load_args, S_PARAM_ST * pparam);
```

File: src/storage/btree_load.c [hunk 8]
```diff
@@ -264,6 +260,40 @@ static int online_index_builder (THREAD_ENTRY * thread_p, BTID_INT * btid_int, H
...
+  OID orig_oid;
+  OID orig_class_oid;
+  MVCC_REC_HEADER orig_mvcc_header;
+} S_PARAM_ST;
+
+static int put_buf_to_record (RECDES * recdes, SORT_ARGS * sort_args, int value_has_null, int key_len,
+			      OID * prev_oid, MVCC_REC_HEADER * mvcc_header, DB_VALUE * dbvalue_ptr, int cur_class,
+			      bool is_btree_ops_log);
+static int get_buf_from_record (OR_BUF * buf, RECDES * recdes, LOAD_ARGS * load_args, S_PARAM_ST * pparam, bool copy);
+static int first_leaf_page_and_init_args (THREAD_ENTRY * thread_p, LOAD_ARGS * load_args, S_PARAM_ST * pparam);
```
Comment (beyondykk9) [inline src/storage/btree_load.c]:
```suggestion
static int get_first_leaf_page_and_init_args (THREAD_ENTRY * thread_p, LOAD_ARGS * load_args, S_PARAM_ST * pparam);
or
static int find_first_leaf_page_and_init_args (THREAD_ENTRY * thread_p, LOAD_ARGS * load_args, S_PARAM_ST * pparam);
```

File: src/storage/btree_load.c [hunk 9]
```diff
@@ -264,6 +260,40 @@ static int online_index_builder (THREAD_ENTRY * thread_p, BTID_INT * btid_int, H
...
+static int make_new_record_on_leaf_page (THREAD_ENTRY * thread_p, LOAD_ARGS * load_args, S_PARAM_ST * pparam,
+					 int *sp_success);
+static int mvcc_delid_invalid (THREAD_ENTRY * thread_p, LOAD_ARGS * load_args, S_PARAM_ST * pparam);
+static int nospace_for_new_oid (THREAD_ENTRY * thread_p, LOAD_ARGS * load_args, int *sp_success);
+static int add_same_key_to_record (THREAD_ENTRY * thread_p, LOAD_ARGS * load_args, S_PARAM_ST * pparam,
+				   int *sp_success);
+static int notify_to_vacuum (THREAD_ENTRY * thread_p, LOAD_ARGS * load_args, S_PARAM_ST * pparam,
+			     char **notify_vacuum_rv_data, char *notify_vacuum_rv_data_bufalign);
+static int heap_scancache_and_attrinfo_start (THREAD_ENTRY * thread_p, SORT_ARGS * args, HEAP_SCANCACHE * scan_cache,
```
Comment (beyondykk9) [inline src/storage/btree_load.c]:
```suggestion
static int heap_scancache_start_for_attrinfo (THREAD_ENTRY * thread_p, SORT_ARGS * args, HEAP_SCANCACHE * scan_cache,
```

File: src/storage/btree_load.c [hunk 10]
```diff
@@ -264,6 +260,40 @@ static int online_index_builder (THREAD_ENTRY * thread_p, BTID_INT * btid_int, H
...
+					      HEAP_CACHE_ATTRINFO * attr_info, int save_cache_last_fix_page);
+static void heap_scancache_and_attrinfo_end (THREAD_ENTRY * thread_p, SORT_ARGS * args, HEAP_SCANCACHE * scan_cache,
```
Comment (beyondykk9) [inline src/storage/btree_load.c]:
```suggestion
static void heap_scancache_end_for_attrinfo (THREAD_ENTRY * thread_p, SORT_ARGS * args, HEAP_SCANCACHE * scan_cache,
```

File: src/storage/page_buffer.h (hunk 2 above)
Comment (ctshim) [reply src/storage/page_buffer.h]: I was mistaken. I'll fix it.

File: src/storage/btree_load.c (hunk 6 above)
Comment (ctshim) [reply src/storage/btree_load.c:277]: That's a good suggestion.
I'll fix it.

File: src/storage/btree_load.c (hunk 5 above)
Comment (ctshim) [reply src/storage/btree_load.c:2408]: I'll add a comment to that function.

File: src/storage/heap_file.c (hunk 3 above)
Comment (ctshim) [reply src/storage/heap_file.c]: Removing an argument from the function heap_attrinfo_read_dbvalues() is not relevant to this issue.

However, it is obvious that it is unnecessary, so I will revise it together.

File: src/storage/btree_load.c (hunk 7 above)
Comment (ctshim) [reply src/storage/btree_load.c]: OK, I'll change it.

File: src/storage/btree_load.c (hunk 8 above)
Comment (ctshim) [reply src/storage/btree_load.c]: OK, I'll change it.

File: src/storage/btree_load.c (hunk 9 above)
Comment (ctshim) [reply src/storage/btree_load.c]: OK, I'll change it.

File: src/storage/btree_load.c (hunk 10 above)
Comment (ctshim) [reply src/storage/btree_load.c]: OK, I'll change it.

File: src/storage/btree_load.c (hunk 4 above)
Comment (ctshim) [reply src/storage/btree_load.c:276]: Temporary structures have been defined to reduce what will be passed to the parameters of the function.
There is no particular meaning to the name of the structure.

File: src/storage/slotted_page.h (hunk 1 above)
Comment (ctshim) [reply src/storage/slotted_page.h]: OK, I will fix it.

Comment (hgryoo) [reply src/storage/slotted_page.h]: Btw, `sizeof(SPAGE_SLOT)` is a fixed size. we use capital characters as code conventions.

e.g) RESERVED_SIZE_IN_PAGE in storage_commmon.c
```
...
#define RESERVED_SIZE_IN_PAGE   (sizeof (FILEIO_PAGE_RESERVED) + sizeof (FILEIO_PAGE_WATERMARK))
...
```

So, they should be
```
#define SPAGE_SLOT_SIZE   (sizeof(SPAGE_SLOT))
#define SPAGE_HEADER_SIZE (sizeof(SPAGE_HEADER))
```

**Review by hyunikn — APPROVED** [review]

**Review by beyondykk9 — APPROVED** [review]

File: src/storage/slotted_page.h (hunk 1 above)
Comment (ctshim) [reply src/storage/slotted_page.h]: I'll fix that, thank you.

**Review by hgryoo — APPROVED** [review]

File: src/storage/btree_load.c [hunk 11]
```diff
@@ -4909,23 +4870,26 @@ online_index_builder (THREAD_ENTRY * thread_p, BTID_INT * btid_int, HFID * hfids
...
-      if (sc == S_ERROR)
-	{
-	  ASSERT_ERROR_AND_SET (ret);
-	  break;
-	}
-      else if (sc == S_END)
-	{
-	  break;
-	}
+      if (sc != S_SUCCESS)
```
Comment (youngjinj) [inline src/storage/btree_load.c:4873]: It is trivial.
Is it to reduce comparison in case of S_SUCCESS?

**Review by youngjinj — APPROVED** [review]

File: src/storage/btree_load.c (hunk 11 above)
Comment (ctshim) [reply src/storage/btree_load.c:4873]: Yes, to reduce comparisons of if statement.

**Review by shparkcubrid — APPROVED** [review]

**Pull request merged**